logstash/lib/logstash/outputs/loki.rb

```diff
@@ -0,0 +1,262 @@
+  config :password, :validate => :string, secret: true, :required => false
+
+  ## 'Client certificate'
+  config :cert, :validate => :path, :required => false
+  config :key, :validate => :path, :required => false
+
+  ## 'TLS'
+  config :ca_cert, :validate => :path, :required => false
+
+  ## 'Loki Tenant ID'
+  config :tenant_id, :validate => :string, :default => "fake", :required => false
+
+  ## 'Maximum number of records to hold in a batch before pushing to Loki'
+  config :batch_size, :validate => :number, :default => 10, :required => false
+
+  ## 'Interval in seconds to wait before pushing a batch of records to Loki'
+  config :batch_wait, :validate => :number, :default => 1, :required => false
+
+  ## 'Array of label names to exclude from sending to Loki'
+  config :exclude_labels, :validate => :array, :default => [], :required => false 
+
+  ## 'Extra labels to add to all log streams'
+  config :external_labels, :valide => :hash,  :default => {}, :required => false
+
+
+  public
+  def register
+    @uri = URI.parse(@url + '/loki/api/v1/push')
+    unless @uri.is_a?(URI::HTTP) || @uri.is_a?(URI::HTTPS)
+      raise LogStash::ConfigurationError, "url parameter must be valid HTTP, currently '#{@url}'"
+    end
+
+    @logger.info("New Loki output", :class => self.class.name)
+
+    @Channel = Concurrent::Channel
+    @batches = {}
+    @entries = @Channel.new
+    @exclude_labels += ["message", "@timestamp"] 
+    @retry_http = 0
+    @delay = 1
+    if ssl_cert?
+      load_ssl
+      validate_ssl_key
+    end
+
+    @Channel.go{run()}
+  end
+
+  def ssl_cert?
+    !@key.nil? && !@cert.nil?
+  end
+
+  def load_ssl
+    @cert = OpenSSL::X509::Certificate.new(File.read(@cert)) if @cert
+    @key = OpenSSL::PKey.read(File.read(@key)) if @key
+  end
+
+  def validate_ssl_key
+    if !@key.is_a?(OpenSSL::PKey::RSA) && !@key.is_a?(OpenSSL::PKey::DSA)
+      raise "Unsupported private key type #{key.class}"
+    end
+  end
+
+  def ssl_opts(uri)
+    opts = {
+      use_ssl: uri.scheme == 'https'
+    }
+
+    if !@cert.nil? && !@key.nil?
+      opts = opts.merge(
+        verify_mode: OpenSSL::SSL::VERIFY_PEER,
+        cert: @cert,
+        key: @key
+      )
+    end
+
+    unless @ca_cert.nil?
+      opts = opts.merge(
+        ca_file: @ca_cert
+      )
+    end
+    opts
+  end 
+
+  def run()
+	  min_wait_checkfrequency = 1/1000 #1 millisecond
+	  max_wait_checkfrequency = @batch_wait / 10
+	  if max_wait_checkfrequency < min_wait_checkfrequency
+		  max_wait_checkfrequency = min_wait_checkfrequency
+    end
+
+    max_wait_check = Concurrent::Channel.tick(max_wait_checkfrequency)
+    loop do
+      Concurrent::Channel.select do |s|
+        s.take(@entries) { |e|
+          if !@batches.has_key?(e.tenant_id)
+            @batches[e.tenant_id] = Batch.new(e)
+            next
+          end
+
+          batch = @batches[e.tenant_id]
+          line = e.entry['line']
+          if batch.size_bytes_after(line) > @batch_size
```

Comment:
I don’t think max size should be in bytes but more in entry count ? Unless you change the parameter name and description

Reply:
It's length of entry(line).

Reply:
We batch per entry in promtail no? And the help description is also saying record

Reply:
No, I didn't mean we batch per entry. We add each entry to the batch and `bytes` is measured by length of line https://github.com/grafana/loki/blob/master/pkg/promtail/client/batch.go#L39

Reply:
To be very frank the whole code I have written is just copy paste of promtail client and batch implementation. It's just ruby version of go code 😜

Reply:
Sorry, now I realized what you said. I will change the description

Reply:
Hey this is great work ! No need to apologize. And you're right it using size so changing the description would do. Also make sure we have a sane default value like promtail:

```
	flags.IntVar(&c.BatchSize, "client.batch-size-bytes", 100*1024, "Maximum batch size to accrue before sending. ")
```

Reply:
Sure